lib/Backend/arm64/LowerMD.cpp

```diff
@@ -791,64 +791,47 @@ LowererMD::GenerateStackProbe(IR::Instr *insertInstr, bool afterProlog)
+        // LDR   r17, [r17, #0]
+        this->CreateAssign(scratchOpnd, IR::IndirOpnd::New(scratchOpnd, 0, TyMachReg, this->m_func), insertInstr);
 
-            }
+        AssertMsg(!IS_CONST_00000FFF(frameSize), "For small size we can just add frameSize to r17");
 
-            IR::Opnd *scratchOpnd2 = IR::RegOpnd::New(nullptr, SP_ALLOC_SCRATCH_REG, TyMachReg, this->m_func);
-            this->CreateAssign(scratchOpnd2, IR::IntConstOpnd::New(frameSize, TyMachReg, this->m_func), insertInstr);
+        IR::Opnd *spAllocRegOpnd = nullptr;
+        IR::Opnd *unusedRegForAlignOpnd = nullptr;
+        if (!afterProlog)
+        {
+            // STP r15,r14,[sp, #-16]
+            spAllocRegOpnd = IR::RegOpnd::New(nullptr, SP_ALLOC_SCRATCH_REG, TyMachReg, this->m_func);
+            unusedRegForAlignOpnd = IR::RegOpnd::New(nullptr, UNUSED_REG_FOR_STACK_ALIGN, TyMachReg, this->m_func);
+            IR::Instr * instrStp = IR::Instr::New(Js::OpCode::STP_PRE,
+                IR::IndirOpnd::New(IR::RegOpnd::New(nullptr, RegSP, TyMachReg, this->m_func), -(2 * MachRegInt), TyMachReg, this->m_func),
+                spAllocRegOpnd, unusedRegForAlignOpnd, this->m_func);
+            insertInstr->InsertBefore(instrStp);
+        }
 
-            instr = IR::Instr::New(Js::OpCode::ADDS, scratchOpnd, scratchOpnd, scratchOpnd2, this->m_func);
-            insertInstr->InsertBefore(instr);
+        // MOV r15, frameSize
+        spAllocRegOpnd = IR::RegOpnd::New(nullptr, SP_ALLOC_SCRATCH_REG, TyMachReg, this->m_func);
+        this->CreateAssign(spAllocRegOpnd, IR::IntConstOpnd::New(frameSize, TyMachReg, this->m_func), insertInstr);
 
-            if (!afterProlog)
-            {
-                Assert(scratchBit.Test(RegEncode[SP_ALLOC_SCRATCH_REG]));
-                opnd = IR::RegBVOpnd::New(scratchBit, TyMachReg, this->m_func);
-                instr = IR::Instr::New(Js::OpCode::LDR, opnd, this->m_func);
-                opnd = IR::IndirOpnd::New(IR::RegOpnd::New(nullptr, RegSP, TyMachReg, this->m_func), (int32)0, TyMachReg, this->m_func);
-                instr->SetSrc1(opnd);
-                insertInstr->InsertBefore(instr);
+        // ADDS r17, r17, r15
+        instr = IR::Instr::New(Js::OpCode::ADDS, scratchOpnd, scratchOpnd, spAllocRegOpnd, this->m_func);
+        insertInstr->InsertBefore(instr);
 
-                IR::RegOpnd* regSpOpnd = IR::RegOpnd::New(nullptr, RegSP, TyMachReg, this->m_func);
-                IR::Instr * instrSub = IR::Instr::New(Js::OpCode::ADD, regSpOpnd, regSpOpnd, IR::IntConstOpnd::New(MachRegInt, TyMachReg, this->m_func), this->m_func);
-                insertInstr->InsertBefore(instrSub);
-            }
+        if (!afterProlog)
+        {
+            // LDP r15,r14,[sp, #16]
+            spAllocRegOpnd = IR::RegOpnd::New(nullptr, SP_ALLOC_SCRATCH_REG, TyMachReg, this->m_func);
+            unusedRegForAlignOpnd = IR::RegOpnd::New(nullptr, UNUSED_REG_FOR_STACK_ALIGN, TyMachReg, this->m_func);
```

Comment:
Can't you re-use the operands from earlier?

Reply:
Removed the code to save/restore x14,x15

---
In reply to: [144412422](https://github.com/Microsoft/ChakraCore/pull/3943#discussion_r144412422) [](ancestors = 144412422)